### ISSUE TITLE
performance / why is it running twice ...

### ISSUE THREAD

**Issue by redaktor** [spencermountain/nlp_compromise#49]
Hey there,
contributing from my fork doesn't make sense because the structure will change to 'only the 3 dictionary files and a factory' soon.
However - let me ask some perfomance questions.
Maybe I missed something, hidden in the code, but
--> several 'autoclosure' functions run every time when a module is required.  

Let's take an example - the conjugation of verbs which is used quite often.
I'll use simple console.log to demonstrate it.

In [src/parents/verb/index](https://github.com/spencermountain/nlp_compromise/blob/master/src/parents/verb/index.js)

put some logs in the conjugate function
```js
the.conjugate = function() {
  console.log( 'BEWARE! conjugate is conjugating' );
  verb_conjugate = require('./conjugate');
  var conjugated = verb_conjugate(the.word);
  console.log( 'conjugate result', conjugated );
  return conjugated; //verb_conjugate(the.word);
}
```
and in the 'autoclosure' form function
```js
the.form = (function() {
	console.log( 'BEWARE! the.form is conjugating' );
	verb_conjugate = require('./conjugate');
	// don't choose infinitive if infinitive == present
	var order = [
	  'past',
	  'present',
	  'gerund',
	  'infinitive'
	];
	var forms = verb_conjugate(the.word);
	console.log( 'forms result', forms );
	for (var i = 0; i < order.length; i++) {
		if (forms[order[i]] === the.word) {
			return order[i];
		}
	}
})()
```

When I do
```
console.log( nlp.verb('last') );
```
it will conjugate

and when I do
```
console.log( nlp.verb('last').conjugate() );
```
it will conjugate twice

**Comment by spencermountain** [comment]
yeah, this is a great point.  the whole '.analysis' thing is poorly implemented. These methods auto-fire because they're used variously around the library, (in places i now forget) and i thought caching it on create was a good idea... it's probably not. The verb one is the worst one, but the other parents do it too

**Comment by spencermountain** [comment]
there was a point when i thought pos.js would just decide 'noun,verb,adj', then `parent.analysis()` would figure out if it was plural, acronym, past-tense, whatever. that kinda fell apart

**Comment by redaktor** [comment]
but it is not about 
> caching it on create
it is about not using the cached from the beginning

the easiest fix:

after
```
the.word = str || '';
```
insert
```js
the.conjugated = {};
```

change 'the.conjugate' to
```js
the.conjugate = function() {
	if (the.conjugated) {
		return the.conjugated;
	}
	verb_conjugate = require('./conjugate');
	return verb_conjugate(the.word);
}
```
and the.form to
```js
the.form = (function() {
	// don't choose infinitive if infinitive == present
	var order = [
		'past',
		'present',
		'gerund',
		'infinitive'
	];
	the.conjugated = verb_conjugate(the.word);
	for (var i = 0; i < order.length; i++) {
		if (the.conjugated[order[i]] === the.word) {
			return order[i];
		}
	}
})()

``

**Comment by redaktor** [comment]
In general it is up to the project using nlp but usually I would rather wait 200ms more when it is loading (before starting to write) if the things done 'on keypress' are much faster ...

**Comment by spencermountain** [comment]
+1 make a pr?

**Comment by spencermountain** [comment]
maybe caching is something we should do for every method.. some kind of window.cache={} or something

**Comment by redaktor** [comment]
Well,
then it would need a constructor/factory as well.
There is no inheritance in this project since the 'index' is a plain object and not a constructor/factory function (I think it is called constructor in AMD and factory in node modules). However - I mean
http://metaduck.com/08-module-pattern-inheritance.html ('module pattern').

And I think basically such a project should have a preprocessor and a constructor/factory.
This is what I want to achieve with the fork. This would be a 'breaking' change. <br>E.g. instead of 
```js
var nlp = require('./nlp');
nlp.do('whatever')
```
it would be 
```js
var nlp = require('./nlp');
nlp().do('whatever');
```

Let's take an example as it is for now :
```
 nlp.verb('go').conjugate();
```
The first require ('nlp.') will "invoke" the 'index.js'.
This will require 'pos.js' and this will require the 'lexicon.js'.
The lexicon conjugates all verbs and declines all adjectives *already* ...
We could cache them –
but wait when I call nlp.verb('go').conjugate() the 'parents/verb/conbjugate' would not know about any other parts of the project (and not about the already cached lexicon) ...

**Comment by redaktor** [comment]
another possibility would be to add a 'singleton' cache module like
```js
module.exports = (function () {
    var cache = (typeof window != 'undefined') ? window.cache : require.cache;
		cache.nlp = {};

    return {
        get: function (key) { return cache.nlp[key]; },
        set: function (key, val) { cache.nlp[key] = val; }
    }
})();
```

**Comment by spencermountain** [comment]
ooh i vote singleton!  will you implement this?

**Comment by spencermountain** [comment]
couple thoughts:
i know hashes hash things, but does it make sense to do this?
``` return cache.nlp[Md5(key)]``` if say, we're caching a long sentence?

lastly, if we're adding a one-liner to every method, we should make sure to write the cache to make this line as cute as possible. something like this?
```nlp.myfunction(str){
    if(nlp.is_cached(str,'myfunction')){return nlp.cached(str,'myfunction')};
}```
or something even nicer, if you dream it up

**Comment by spencermountain** [comment]
it would be annoying if our cache-hit line became hard to maintain

how much faster do you think it would be, for our tests and things?

**Comment by spencermountain** [comment]
(last thing) ..... or maybe it could *all* be implimented in index.js?
is that where you were going?
ie, it just gets inbetween nlp and it's functions.. invisibly.

**Comment by redaktor** [comment]
Yes - what I did is similar. It reads e.g.
```js
nlp.myfunction(str) {
  var cached = cache.get(str, 'myFn');
  if (cached) { return cached; }
}
```

**Comment by redaktor** [comment]
a comparison (not ready yet) ...

| test  | redaktor  | original |
| ----- | --------- | -------- |
| ✔ ambiguous contractions  | 19ms  | 27ms  |
| ✔ referenced by  | 14ms  | 30ms  |
| ✔ reference to  | 7ms  | 17ms  |
| ✔ referables  | 2ms  | 4ms  |
| ✔ noun.pronoun  | 1ms  | 6ms  |
| ✔ sentence.people  | 4ms  | 8ms  |
| ✔ phrasal verbs  | 4ms  | 6ms  |
| ✔ word tokenization  | 1ms  | 2ms  |
| ✔ is_person  | 1ms  | 5ms  |
| ✔ noun.article  | 1ms  | 2ms  |
| ✔ conjugate future  | 2ms  | 5ms  |
| ✔ conjugate present  | 2ms  | 5ms  |
| ✔ conjugate past  | 3ms  | 6ms  |
| ✔ verb conjugate  | 12ms  | 33ms  |
| ✔ negation  | 16ms  | 32ms  |
| ✔ to_comparative  | 1ms  | 2ms  |
| ✔ to_superlative  | 0ms  | 0ms  |
| ✔ to_adverb  | 3ms  | 10ms  |
| ✔ adverb.to_adjective  | 4ms  | 4ms  |
| ✔ sentences  | 2ms  | 2ms  |
| ✔ is_plural  | 0ms  | 3ms  |
| ✔ pluralize  | 3ms  | 18ms  |
| ✔ singularize  | 3ms  | 15ms  |
| ✔ ngram  | 2ms  | 2ms  |
| ✔ adj_to_noun  | 0ms  | 1ms  |
| ✔ dates  | 18ms  | 25ms  |
| ✔ americanize  | 1ms  | 1ms  |
| ✔ britishize  | 0ms  | 0ms  |
| ✔ syllables  | 3ms  | 2ms  |
| ✔ unicode  | 0ms  | 0ms  |
| ✔ value  | 6ms  | 11ms  |
| ✔ pos  | 22ms  | 38ms  |
| ✔ spot  | 11ms  | 17ms  |

**Comment by spencermountain** [comment]
WOW!

**Comment by redaktor** [comment]
just btw:
why not make use of factories? 
Functions returning functions totally avoid the duplication of code known as "spaghetti code" ;)
An example: src/section.js – 
it has now **109** lines but it is the same like this **57** lines (52% code) :

```js
// a section is a block of text, with an arbitrary number of sentences
// these methods are just wrappers around the ones in sentence.js

exports.sugar = function(type, todo) {
	if (todo === 'reduce') {
		return function(options) {
			return this.sentences.map(function(s) {
				return s[type](options);
			}).reduce(function(arr, a) {
				return arr.concat(a);
			}, []);
		}
	} else if (todo === 'transform') {
		return function() {
			this.sentences = this.sentences.map(function(s) {
				return s[type]();
			});
			return this;
		}
	} else {
		return function() {
			return this.sentences.map(function(s) {
				return s[type]();
			});
		}
	}
}

exports.Section = function(sentences) {
  //var the = this;
  this.sentences = sentences || [];

  this.text = function() {
    return this.sentences.map(function(s) {
      return s.text();
    }).join(' ');
  }
  this.tense = exports.sugar('tense');

  // pluck out wanted data from sentences
  this.tags = exports.sugar('tags');
  this.entities = exports.sugar('entities', 'reduce');
  this.people = exports.sugar('people', 'reduce');
  this.nouns = exports.sugar('nouns', 'reduce');
  this.adjectives = exports.sugar('adjectives', 'reduce');
  this.verbs = exports.sugar('verbs', 'reduce');
  this.adverbs = exports.sugar('adverbs', 'reduce');
  this.values = exports.sugar('values', 'reduce');
	
  // transform the sentences
  this.negate = exports.sugar('negate', 'transform');
  this.to_past = exports.sugar('to_past', 'transform');
  this.to_present = exports.sugar('to_present', 'transform');
  this.to_future = exports.sugar('to_future', 'transform');

}
module.exports = exports.Section;
```

**Comment by spencermountain** [comment]
i like this too. feel free to add it. nicely done.

**Comment by redaktor** [comment]
I pushed it to the fork (renamed exports.sugar to exports.fn for pretending to be serious). 

It will maybe be next weekend when I can consider the fork as not being broken anywhere
and then you'll get a clue of the whole thing.
We are merging some ideas with the redaktor translation center where anyone can propose new translations via web and things go to git automatically (to our fork first). It is similar to the Adobe Translation Center. It would be sweet if you could test your rules and regexes there directly in a fabulous editor. There are many things in the future I'll do anyway (we need this too).
As said it is WIP.

Now as said persistent decoupling of logic and shared data will further
- prevent duplication 
- speed things up

Questions:
How do we deal with the maximum size of the cache (for now AKA memory ;) ...
Set it in the options ?
How big would a cache of the bible pos be and what would we do if it is bigger than the memory?
What would Jesus do?
My idea is to implement a cache.lock done after the factory (for the important lexicon words).
They should never be deleted. And then?

**Comment by spencermountain** [comment]
;) i'm certain jesus gave up on javascript a long time ago.

if the average vocabulary is 50k words, and each word is 5 bytes, you'd have still room left on a floppy drive, if we only cache-by-word. 
that sounds like a tricky problem.

**Comment by redaktor** [comment]
:+1: finished pushing just now. coding again.

**Comment by redaktor** [comment]
ah - I see -
[stats](http://www.biblebelievers.com/believers-org/kjv-stats.html) = 788k * 5  = 3940k ...

But the problem is that we are caching the underlying methods seperately and some have object keys as well.
Let's say the average caching hit is 7 and each word has at least + a key (2) and if we store objects it is *3

3940k * 7 * 2 * 3 = 165480k = 158MB.

Well - it doesn't sound that worse.
My Amiga 4000 would fail but the inexistent Apple Watch would succeed.
Basically I would set 256MB as a max. in options for now and disable caching if that is reached (by the typical performant "assumption" size checking, which could even run asynchronous [doesn't matter if it is assumption anyway]).
That is easy and other things can be done

**Comment by redaktor** [comment]
please also note that browserify currently does not work in the fork because I did not code the splitting of .min and normal files and cleaning up. Doing this now.

**Comment by redaktor** [comment]
@spencermountain I am attaching this beacause we talked about performance here and I was pretty active today (pushing tomorrow)  and now I reached 'verb/conjugate' with the comment
```
// this method is the slowest in the whole library, basically TODO:whaaa
function predict(w) {
	var endsWith = function(str, suffix) {
		return str.indexOf(suffix, str.length - suffix.length) !== -1;
	}
	var arr = Object.keys(suffixes);
	for (i = 0; i < arr.length; i++) {
		if (endsWith(w, arr[i])) {
			return suffixes[arr[i]];
		}
	}
	return 'infinitive';
}
```
It is easy to fix:
A suffix is the last 1-4 chars of the word, so it would be much faster just to check
```js
function predict(w) {
  for (var i = -4; i < 0; i++) {
    var sliced = w.slice(i);
    if (suffixes.hasOwnProperty(sliced)) { return suffixes[sliced]; }
  }
  return 'infinitive';
}
```
For maximal flexibility:
Check the maximum length of the suffix keys. In the fork length will already be delivered by data module...

**Comment by spencermountain** [comment]
BONANZA!

**Comment by redaktor** [comment]
hm - still working. Wanna have all /parents complete. Just btw - just found this snippet in reference_to :
```js
// reverse the terms to loop through backward..
interested = interested.reverse()
for(var i=0; i < interested.length; i++){ '...' }
```
I would call this a bit of an anti-pattern. It is easier to do just
```js
var i = interested.length;
while(i--) { '...' }
```
This is doing the same ;)

**Comment by spencermountain** [comment]
hey Sebastian, am starting a 2.0 rewrite,and will include these awesome ideas in that branch.

**Comment by redaktor** [comment]
ok - cool.
had three hard days (ycombinator deadline) - feel free to compare the files in the fork and "cherry pick".
The 'values' are also close to complete and can now parse all the worlds measurement units (completely understandable for machines putting quantities in context). Let you know.
